### ISSUE TITLE
Outline revised curriculum

### ISSUE THREAD

**Issue by bradleyboehmke** [uc-python/advanced-python-datasci#16]
We have decided to modify the proposed outline and curriculum to be more of an end-to-end [machine learning / data science] project development with 50% of time spent on the machine learning aspects and 50% of time spent on the technology and programming that goes with it.

We envision the workshop to expose people to the machine learning workflow along with required technology topics to implement the workflow in a typical project. Most importantly, the students will implement these workflows so they walk away from the workshop with a fully built out project.

**This issues serves as a place for us to propose and finalize what this new curriculum/agenda should look like.**

**Comment by eswan18** [comment]
Along with the schedule, something we should also decide is what data set we want to work with. We use the planes data for the first two classes, and it's already pretty clean, so I don't think that's a good candidate.

One site I discovered recently has a catalog of ML datasets, so it might be worth looking over: http://archive.ics.uci.edu/ml/datasets.php

**Comment by bradleyboehmke** [comment]
The UCI repo has been a classic for some time. A couple other considerations:

- I have some content already built that leverages the [Ames, IA real estate data set](https://bradleyboehmke.github.io/HOML/intro.html#data). We used it in our [HOML book](https://bradleyboehmke.github.io/HOML/) and I have translated some of it over to sklearn.
- We also could use the [84.51 Complete Journey public data set](https://bradleyboehmke.github.io/completejourney/). I turned this into an R package a couple years back and Jay turned it into a [Python package](https://pypi.org/project/completejourney-py/), which makes it easily importable. 

I was also thinking about trying to find a continuous realtime feed data source. The end-to-end project could be creating an program that runs daily and gets the new data daily, preprocesses it, checks for significant changes in data distribution, and then runs the model.

**Comment by eswan18** [comment]
I guess my opinion really depends on how much you already have prepared around the Ames dataset. Will that make it significantly easier to make the sections on modeling? If so, for simplicity that seems like the best option – though it's really up to you, because I think the time savings would mainly be in preparing your sections.

If not Ames, the Complete Journey seems like a great option. I was totally unaware of it somehow, but we know the schema inside and out, and it's already clean – so we don't need to spend a lot of time on wrangling, which was already covered thoroughly in the first two classes.

While I like the realtime idea, I think it presents a lot of additional complexity. Students wouldn't get to see the "daily" nature of it much during the course itself, and running something daily on a laptop that could be off or sleeping is tricky (plus complexity around scheduling on different OSes). The cloud would be a good option for that but is an enormous can of worms.

**In short**: I think we should use either Ames or the Complete Journey; the former if it gives us a significant headstart on content prep, otherwise the latter, and I'd defer to your judgment on that.

**Comment by bradleyboehmke** [comment]
For the ames data Brandon and I used it extensively in [our book](https://bradleyboehmke.github.io/HOML/) and in our [ML with R workshop](https://github.com/uc-r/Advanced-R). I have already translated some of this content over to using Python and scikit-learn but not all of it. But having this content lets us understand what we should expect when translating other parts of the content from R to Python.

With that being said, the complete journey content would be fun considering it is newer but I honestly don't know what to expect as far as predictive performance in models so it would take some playing with to make sure I have my arms around the data and how well models work with it.  With that being said, it would be truly original content that could open the door for other opportunities including bringing this workshop in-house 🤷‍♂️.

**Comment by eswan18** [comment]
I think ultimately it's your call – whichever we choose won't make a big difference in the technical aspects (Git, project structure, coding style). If you want to tackle the 84.51˚ data, I'm absolutely game, but if you would prefer to take advantage of the Ames material you already have then that makes perfect sense to me. You can think it over if you want and let me know.

**Comment by bradleyboehmke** [comment]
Proposed agenda & division of labor:

## Ethan

- Git and GitHub – Making a repo, add-commit-push workflow
- Laying out a Python project in files and folders
- How Python imports work within a project directory
- Managing dependencies using requirements.txt or environment.yaml files, with some callbacks to the Environments section from the intermediate class
- Modularity and good coding – thinking in functions, code that does one thing and does it well
- If time permitted, I could also touch on the basics of writing tests, though I’d have to think more about whether there’s a satisfying way to discuss tests in just an hour for people who have never written them before.

## Brad 

@eswan18, I made a few slight revisions from your email

- Model training & selection - slight revision of intermediate section; shows the typical workflow of model training
- Feature engineering - how to work with different types of features
- Model tracking & registration - discussing why we track and version models; illustrate with MLflow
- Model deployment & monitoring - discussion around ways to deploy models and also understanding drift

## Agenda

- Day 1
   - Git and version control (Ethan)
   - Model training & selection (Brad)
- Day 2
   - Modular code part 1 (Ethan)
   - Feature engineering (Brad)
- Day 3
   - Model tracking & registration (Brad)
   - Modular code part 2 (and potentially unit tests) (Ethan)
- Day 4
   - Model deployment & monitoring (Brad)
   - “Reproducibility” with dependency tracking (Ethan)

**Comment by eswan18** [comment]
I'm going to close this @bradleyboehmke – but if you think there's more to do yet, feel free to reopen.
